### PR TITLE
MANTA-5285 rebalancer could do a better job of tracking shark available MB

### PR DESCRIPTION
**Pull request by rjloura**
MANTA-5307 shark assignment generator thread could lose objects that put the assignment over max_size
MANTA-5297 rebalancer EvacuateJob structure should have a config field

### CODE REVIEW

**Comment by rjloura** [comment]
I'm posting this as a draft for now.  There are still some changes required to the test code.  
A considerable amount of the changes are refactors and additional tests.  Once the tests have been updated I will annotate the changes.

File: manager/src/metrics.rs [hunk 1]
```diff
@@ -70,11 +70,14 @@ pub fn metrics_init(cfg: metrics::ConfigMetrics) {
...
-    let labels: HashMap<String, String> =
-        match metrics::get_const_labels().lock().unwrap().clone() {
-            Some(cl) => cl,
-            None => HashMap::new(),
-        };
+    let labels: HashMap<String, String> = match metrics::get_const_labels()
+        .lock()
+        .expect("metrics const labels")
+        .clone()
+    {
```
Comment (kellymclaughlin) [inline manager/src/metrics.rs:77]: The use of `match` together with the chained function calls ends up a bit hard to follow for me. One option I think  is a bit cleaner is:
```
  let labels: HashMap<String, String> = metrics::get_const_labels()
        .lock()
        .expect("metrics const labels")
        .clone()
        .unwrap_or_else(|| HashMap::new());
```

File: manager/src/metrics.rs [hunk 2]
```diff
@@ -70,11 +70,14 @@ pub fn metrics_init(cfg: metrics::ConfigMetrics) {
...
     let mut metrics = metrics::register_metrics(&cfg);
 
-    let labels: HashMap<String, String> =
-        match metrics::get_const_labels().lock().unwrap().clone() {
-            Some(cl) => cl,
-            None => HashMap::new(),
-        };
+    let labels: HashMap<String, String> = match metrics::get_const_labels()
+        .lock()
+        .expect("metrics const labels")
```
[the remark this answers is not in the document]
Comment (kellymclaughlin) [reply manager/src/metrics.rs:75]: Good call!

File: manager/src/jobs/evacuate.rs [hunk 3]
```diff
@@ -2065,60 +2311,215 @@ fn _channel_send_assignment(
...
+                );
+                job_action.mark_object_error(
+                    &eobj.id,
+                    EvacuateObjectError::BadMantaObject,
+                );
+                return Err(AssignmentAddObjectError::BadMantaObject);
+            }
+        };
+
+    let source = match manta_object
```
Comment (kellymclaughlin) [inline manager/src/jobs/evacuate.rs:2390]: This is another case where I think using `unwrap_or_else` instead of the `match` would improve the clarity.

File: manager/src/jobs/evacuate.rs [hunk 4]
```diff
@@ -3326,6 +3724,186 @@ mod tests {
...
+        let mut g = StdThreadGen::new(10);
+        let mut test_objects = HashMap::new();
+        let num_objects = 100;
+        for _ in 0..num_objects {
+            let mut mobj = MantaObject::arbitrary(&mut g);
+            let mut sharks = vec![];
+
+            mobj.content_length = 1024 * 1024; // 1MiB
+                                               // first pass: 1 or 2
+                                               // second pass: 3 or 4
```
[the remark this answers is not in the document]
Comment (rjloura) [reply manager/src/jobs/evacuate.rs:3811]: done

File: manager/src/jobs/evacuate.rs [hunk 5]
```diff
@@ -75,12 +76,13 @@ use diesel::sql_types;
...
 // the fields in 'struct EvacuateObject'
 table! {
-    use diesel::sql_types::{Text, Integer, Nullable, Jsonb};
+    use diesel::sql_types::{BigInt, Text, Integer, Nullable, Jsonb};
     evacuateobjects (id) {
         id -> Text,
         assignment_id -> Text,
         object -> Jsonb,
         shard -> Integer,
+        size -> BigInt,
```
[the remark this answers is not in the document]
Comment (rjloura) [reply manager/src/jobs/evacuate.rs]: done

File: manager/src/metrics.rs (hunk 1 above)
Comment (rjloura) [reply manager/src/metrics.rs:77]: done

File: manager/src/config.rs [hunk 6]
```diff
@@ -267,21 +274,18 @@ fn _config_update_signal_handler(
 #[cfg(test)]
 mod tests {
     use super::*;
+    use crate::test_util::{
```
[the remark this answers is not in the document]
Comment (rjloura) [reply manager/src/config.rs:277]: yeah probably just do `use crate::test_util` and reference it everywhere.

**Comment by rjloura** [comment]
> I'm seeing the following test failure:
> 
> ```
> 11:18:00:manager(MANTA-5285_better_available_mb_checking) $ cargo test jobs::evacuate::tests::assignment_processing_test -- --test-threads=1 
>     Finished test [unoptimized + debuginfo] target(s) in 0.15s
>      Running /home/kelly/joyent/repos/manta-rebalancer/target/debug/deps/manager-f872b11c6a5459f1
> 
> running 1 test
> test jobs::evacuate::tests::assignment_processing_test ... error: test failed, to rerun pass '--lib'
> ```
> 
> Do you see this failure too?

I think this is due to running the tests in (default) multi-threaded mode.  If you do `--test-threads=1` does that fix it?

File: manager/src/jobs/evacuate.rs [hunk 7]
```diff
@@ -3456,6 +3805,288 @@ mod tests {
         }
     }
 
+    #[test]
+    fn calculate_available_mb_test() {
```
[the remark this answers is not in the document]
Comment (rjloura) [reply manager/src/jobs/evacuate.rs:3809]: added some more tests.

**Comment by kellymclaughlin** [comment]
> > I'm seeing the following test failure:
> > ```
> > 11:18:00:manager(MANTA-5285_better_available_mb_checking) $ cargo test jobs::evacuate::tests::assignment_processing_test -- --test-threads=1 
> >     Finished test [unoptimized + debuginfo] target(s) in 0.15s
> >      Running /home/kelly/joyent/repos/manta-rebalancer/target/debug/deps/manager-f872b11c6a5459f1
> > 
> > running 1 test
> > test jobs::evacuate::tests::assignment_processing_test ... error: test failed, to rerun pass '--lib'
> > ```
> > 
> > 
> > Do you see this failure too?
> 
> I think this is due to running the tests in (default) multi-threaded mode. If you do `--test-threads=1` does that fix it?

No, I did run it with `--test-thread=1`

**Comment by kellymclaughlin** [comment]
> > > I'm seeing the following test failure:
> > > ```
> > > 11:18:00:manager(MANTA-5285_better_available_mb_checking) $ cargo test jobs::evacuate::tests::assignment_processing_test -- --test-threads=1 
> > >     Finished test [unoptimized + debuginfo] target(s) in 0.15s
> > >      Running /home/kelly/joyent/repos/manta-rebalancer/target/debug/deps/manager-f872b11c6a5459f1
> > > 
> > > running 1 test
> > > test jobs::evacuate::tests::assignment_processing_test ... error: test failed, to rerun pass '--lib'
> > > ```
> > > 
> > > 
> > > Do you see this failure too?
> > 
> > 
> > I think this is due to running the tests in (default) multi-threaded mode. If you do `--test-threads=1` does that fix it?
> 
> No, I did run it with `--test-thread=1`

Does the test pass for you? If it's just some environment issue on my end then no need to hold this up for that.

**Comment by rjloura** [comment]
> > > > I'm seeing the following test failure:
> > > > ```
> > > > 11:18:00:manager(MANTA-5285_better_available_mb_checking) $ cargo test jobs::evacuate::tests::assignment_processing_test -- --test-threads=1 
> > > >     Finished test [unoptimized + debuginfo] target(s) in 0.15s
> > > >      Running /home/kelly/joyent/repos/manta-rebalancer/target/debug/deps/manager-f872b11c6a5459f1
> > > > 
> > > > running 1 test
> > > > test jobs::evacuate::tests::assignment_processing_test ... error: test failed, to rerun pass '--lib'
> > > > ```
> > > > 
> > > > 
> > > > Do you see this failure too?
> > > 
> > > 
> > > I think this is due to running the tests in (default) multi-threaded mode. If you do `--test-threads=1` does that fix it?
> > 
> > 
> > No, I did run it with `--test-thread=1`
> 
> Does the test pass for you? If it's just some environment issue on my end then no need to hold this up for that.

Looks good on my end.  Double check that you are running with `--test-threads=1` (plural).  You can also try `make jobtests`.  
```
rui@hayek:~/git/rebalancer$ cargo test jobs::evacuate::tests::assignment_processing_test |grep 'test result'
    Finished test [unoptimized + debuginfo] target(s) in 0.25s
     Running target/debug/deps/rebalancer_agent-0c13808520bdbafa
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 6 filtered out
     Running target/debug/deps/manager-f872b11c6a5459f1
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 18 filtered out
     Running target/debug/deps/rebalancer_adm-d69632147a1e58fe
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 5 filtered out
     Running target/debug/deps/rebalancer_manager-58423909e1d1053a
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 3 filtered out
     Running target/debug/deps/rebalancer-6c2b27db37cb6dfa
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
rui@hayek:~/git/rebalancer$ 
```
